node/core/provisioner/src/disputes/mod.rs

```diff
@@ -0,0 +1,454 @@
+	for (session_index, candidate_hash, _) in active {
+		match onchain.get(&(session_index, candidate_hash)) {
+			Some(d) => {
+				// handle double voting here - if a validator has voted both `for` and `against` at the same time for a dispute - ignore its vote
+				// XOR-ing `for` and `against` gives exactly that
+				if (d.validators_for.clone() ^ &d.validators_against).count_ones() >=
+					supermajority_threshold(d.validators_for.len())
+				{
+					partitioned.can_conclude_onchain.push((session_index, candidate_hash));
+				} else {
+					partitioned.cant_conclude_onchain.push((session_index, candidate_hash));
+				}
+			},
+			None => partitioned.unknown_onchain.push((session_index, candidate_hash)),
+		};
+	}
+
+	// TODO: Prioritise the votes in each partition. Is it still necessary?
+
+	// ... and CONCLUDED in two more
+	for (session_index, candidate_hash, _) in concluded {
+		match onchain.get(&(session_index, candidate_hash)) {
+			Some(_) => partitioned.concluded_known_onchain.push((session_index, candidate_hash)),
+			None => partitioned.concluded_unknown_onchain.push((session_index, candidate_hash)),
+		}
+	}
+
+	partitioned
+}
+
+// Helper trait to obtain the value of vote for `InvalidDisputeStatementKind` and `ValidDisputeStatementKind`.
+// The alternative was to pass a bool to `fn should_keep_vote` explicitly but it's pointless as the value is already 'encoded' in the type.
+trait VoteType {
+	fn vote_value() -> bool;
+}
+
+impl VoteType for InvalidDisputeStatementKind {
+	fn vote_value() -> bool {
+		false
+	}
+}
+
+impl VoteType for ValidDisputeStatementKind {
+	fn vote_value() -> bool {
+		true
+	}
+}
+
+fn should_keep_vote<T: VoteType>(
+	v: &(T, ValidatorIndex, ValidatorSignature),
+	onchain_state: &DisputeState,
+) -> bool {
+	let local_vote = T::vote_value();
+	let (_, validator_index, _) = v;
+	let in_validators_for = *onchain_state
+		.validators_for
+		.get(validator_index.0 as usize)
+		.as_deref()
+		.unwrap_or(&false);
+	let in_validators_against = *onchain_state
+		.validators_against
+		.get(validator_index.0 as usize)
+		.as_deref()
+		.unwrap_or(&false);
+
+	if in_validators_for && in_validators_against {
+		// The validator has double voted and runtime knows about this. Ignore this vote.
+		return false
+	}
+
+	if local_vote && in_validators_against || !local_vote && in_validators_for {
+		// local vote differs from the onchain vote
+		// we need this vote to punish the offending validator
+		return true
+	}
+
+	// The vote is valid. Return true if it is not seen onchain.
+	!(in_validators_for || in_validators_against)
+}
+
+// Entrypoint to code relying on staging runtime API.
+// If the staging API is not enabled - returns an empty HashMap.
+async fn get_onchain_disputes<Sender>(
+	_sender: &mut Sender,
+	_relay_parent: Hash,
+) -> Result<HashMap<(SessionIndex, CandidateHash), DisputeState>, GetOnchainDisputesError>
+where
+	Sender: overseer::ProvisionerSenderTrait,
+{
+	let _onchain = Result::<
```

Comment:
nit: I'd prefer to use an explicit `let _ = foo;` for the `#[cfg(not(feature = "staging-client"))]`-case rather than having prefixed all identifiers with `_`.

Reply:
Fixed.